posthog/temporal/workflows/backfill_batch_export.py

```diff
@@ -0,0 +1,363 @@
+import asyncio
+import collections.abc
+import dataclasses
+import datetime as dt
+import json
+import typing
+
+import temporalio
+import temporalio.activity
+import temporalio.client
+import temporalio.common
+import temporalio.exceptions
+import temporalio.workflow
+from django.conf import settings
+
+from posthog.batch_exports.service import BackfillBatchExportInputs
+from posthog.temporal.client import connect
+from posthog.temporal.workflows.base import PostHogWorkflow
+from posthog.temporal.workflows.batch_exports import (
+    CreateBatchExportBackfillInputs,
+    UpdateBatchExportBackfillStatusInputs,
+    create_batch_export_backfill_model,
+    get_batch_exports_logger,
+    update_batch_export_backfill_model_status,
+)
+
+
+class HeartbeatDetails(typing.NamedTuple):
+    """Details sent over in a Temporal Activity heartbeat."""
+
+    schedule_id: str
+    start_at: str
+    end_at: str
+    wait_start_at: str
+
+    def make_activity_heartbeat_while_running(
+        self, function_to_run: collections.abc.Callable, heartbeat_timeout: dt.timedelta, factor: int = 2
+    ) -> collections.abc.Callable[..., collections.abc.Coroutine]:
+        """Return a callable that returns a coroutine that hearbeats with these HeartbeatDetails.
+
+        The returned callable wraps 'function_to_run' while heartbeatting 'factor' times for every
+        'heartbeat_timeout'.
+        """
+
+        async def heartbeat() -> None:
+            """Heartbeat factor times every heartbeat_timeout."""
```

Comment:
I'm confused, it seems like this sleeps for `timeout / factor` and then heartbeats once and returns? Should this recur or loop? (I'm probably missing something obvious.)

Reply:
Ah yeah, this is missing a `while` loop, I'll add it in my next commit. Thank you!

Without the loop Python will run the task until completion whenever we `await`, effectively heart-beating only once. This is not what we want, as we may be waiting for a while until a backfill finishes, and we will certainly hit the `heartbeat_timeout` if we only heartbeat once.

I'm not super happy with the way we currently handle heartbeats: There is potential for us to abstract these so that we can re-use them in all workflows (and avoid me forgetting while loops!). Every workflow should implement heartbeats to allow cancellation and recovery, I just haven't wrapped my head around at what the heartbeat abstraction should look like; some Temporal samples use decorators which is effectively what I'm doing here...

Anyways, something to tackle in a follow-up PR.